src/gennylib/include/gennylib/topology.hpp

```diff
@@ -0,0 +1,211 @@
+ * to a Topology object to execute.
+ *
+ * The idea is to create a visitor that focuses on each part of the cluster,
+ * allow the Topology object to handle traversal, and keep application-level
+ * code as topology-agnostic as possible.
+ */
+class TopologyVisitor {
+public:
+    virtual void onBeforeTopology(const TopologyDescription&) {}
+    virtual void onAfterTopology(const TopologyDescription&) {}
+
+    virtual void onMongod(const MongodDescription&) {}
+    virtual void onMongos(const MongosDescription&) {}
+
+    virtual void onBeforeReplSet(const ReplSetDescription&) {}
+    virtual void onAfterReplSet(const ReplSetDescription&) {}
+
+    virtual void onBeforeSharded(const ShardedDescription&) {}
+    virtual void onAfterSharded(const ShardedDescription&) {}
+
+
+    /** Misc hooks that most visitors won't need. **/
+
+    // Called between mongods in a replica set.
+    virtual void onBetweenMongods(const ReplSetDescription&) {}
+
+    // Called before/after/between visiting shards.
+    virtual void onBeforeShards(const ShardedDescription&) {}
+    virtual void onAfterShards(const ShardedDescription&) {}
+    virtual void onBetweenShards(const ShardedDescription&) {}
+
+    // Called before/after/between visiting mongoses.
+    virtual void onBeforeMongoses(const ShardedDescription&) {}
+    virtual void onAfterMongoses(const ShardedDescription&) {}
+    virtual void onBetweenMongoses(const ShardedDescription&) {}
+
+    virtual ~TopologyVisitor() = 0;
+};
+
+// Be careful changing the traversal order of the cluster; visitors may depend on it.
+struct TopologyDescription {
+    virtual void accept(TopologyVisitor&) = 0;
+    virtual ~TopologyDescription() = 0;
+};
+
+struct MongodDescription : public TopologyDescription {
+    std::string mongodUri;
+
+    void accept(TopologyVisitor& v) { v.onMongod(*this); }
+};
+
+struct MongosDescription : public TopologyDescription {
+    std::string mongosUri;
+
+    void accept(TopologyVisitor& v) { v.onMongos(*this); }
+};
+
+struct ReplSetDescription : public TopologyDescription {
+    std::string primaryUri;
+    bool configsvr = false;
+    std::vector<MongodDescription> nodes;
+
+    void accept(TopologyVisitor& v) { 
+        v.onBeforeReplSet(*this); 
+        for (int i = 0; i < nodes.size() - 1; i++) {
+            nodes[i].accept(v);
+            v.onBetweenMongods(*this);
+        }
+        nodes[nodes.size() - 1].accept(v);
+        v.onAfterReplSet(*this); 
+    }
+};
+
+struct ShardedDescription : public TopologyDescription {
+    ReplSetDescription configsvr;
+    std::vector<ReplSetDescription> shards;
+    std::vector<MongosDescription> mongoses;
+
+    void accept(TopologyVisitor& v) { 
+        v.onBeforeSharded(*this); 
+        configsvr.accept(v);
+
+        v.onBeforeShards(*this); 
+        for (int i = 0; i < shards.size() - 1; i++) {
+            shards[i].accept(v);
+            v.onBetweenShards(*this);
+        }
```

Comment:
This isn't possible with the `onBetweenShards`, since we don't want it called after the last iteration. Maybe a bit of an unnecessary visit method, but `ToJsonVisitor` used it and it was nice for debugging.